<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Подключение-библиотеки" data-toc-modified-id="Подключение-библиотеки-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подключение библиотеки</a></span></li><li><span><a href="#Загрузим-данные" data-toc-modified-id="Загрузим-данные-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузим данные</a></span></li><li><span><a href="#Изучим-данные" data-toc-modified-id="Изучим-данные-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Изучим данные</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Предобработка данных</a></span></li></ul></li><li><span><a href="#Анализ" data-toc-modified-id="Анализ-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ</a></span><ul class="toc-item"><li><span><a href="#Первый-график-временного-ряда" data-toc-modified-id="Первый-график-временного-ряда-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Первый график временного ряда</a></span></li><li><span><a href="#График-со-скользящими-средними" data-toc-modified-id="График-со-скользящими-средними-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>График со скользящими средними</a></span></li><li><span><a href="#Исследуем-стационарные-ряды" data-toc-modified-id="Исследуем-стационарные-ряды-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Исследуем стационарные ряды</a></span></li><li><span><a href="#Тренды-и-сезонность" data-toc-modified-id="Тренды-и-сезонность-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Тренды и сезонность</a></span></li><li><span><a href="#Проверка-на-дисперсию-математическим-способом" data-toc-modified-id="Проверка-на-дисперсию-математическим-способом-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Проверка на дисперсию математическим способом</a></span></li><li><span><a href="#Проверка-на-станционарность-математическим-способом" data-toc-modified-id="Проверка-на-станционарность-математическим-способом-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Проверка на станционарность математическим способом</a></span><ul class="toc-item"><li><span><a href="#Тест-Дики-Фуллера" data-toc-modified-id="Тест-Дики-Фуллера-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>Тест Дики-Фуллера</a></span></li><li><span><a href="#Тест-KPSS---Квятковский-Филлипс-Шмидт-ШинаДики-Фуллер" data-toc-modified-id="Тест-KPSS---Квятковский-Филлипс-Шмидт-ШинаДики-Фуллер-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>Тест KPSS - Квятковский-Филлипс-Шмидт-ШинаДики-Фуллер</a></span></li></ul></li></ul></li><li><span><a href="#Обучение-и-тестирование" data-toc-modified-id="Обучение-и-тестирование-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение и тестирование</a></span><ul class="toc-item"><li><span><a href="#Деление-на-обучающей-и-тестовой-выборки" data-toc-modified-id="Деление-на-обучающей-и-тестовой-выборки-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Деление на обучающей и тестовой выборки</a></span></li><li><span><a href="#LinearRegression" data-toc-modified-id="LinearRegression-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>LinearRegression</a></span></li><li><span><a href="#DecisionTreeRegressor" data-toc-modified-id="DecisionTreeRegressor-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>DecisionTreeRegressor</a></span></li><li><span><a href="#RandomForestRegressor" data-toc-modified-id="RandomForestRegressor-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>RandomForestRegressor</a></span></li><li><span><a href="#LightGBMRegressor" data-toc-modified-id="LightGBMRegressor-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>LightGBMRegressor</a></span></li><li><span><a href="#CatBoostRegressor" data-toc-modified-id="CatBoostRegressor-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>CatBoostRegressor</a></span></li><li><span><a href="#Результаты" data-toc-modified-id="Результаты-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Результаты</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

#  Прогнозирование заказов такси

Компания «Чётенькое такси» собрала исторические данные о заказах такси в аэропортах. Чтобы привлекать больше водителей в период пиковой нагрузки, нужно спрогнозировать количество заказов такси на следующий час. Постройте модель для такого предсказания.

Значение метрики *RMSE* на тестовой выборке должно быть не больше 48.

Вам нужно:

1. Загрузить данные и выполнить их ресемплирование по одному часу.
2. Проанализировать данные.
3. Обучить разные модели с различными гиперпараметрами. Сделать тестовую выборку размером 10% от исходных данных.
4. Проверить данные на тестовой выборке и сделать выводы.


Данные лежат в файле `taxi.csv`. Количество заказов находится в столбце `num_orders` (от англ. *number of orders*, «число заказов»).

## Подготовка

### Подключение библиотеки

In [25]:
!pip install -q jupyternotify

In [26]:
# Уведомление о завершение работы определенного ячейка (очень пригодится для машинного обучения)
import jupyternotify
%load_ext jupyternotify

The jupyternotify extension is already loaded. To reload it, use:
  %reload_ext jupyternotify


In [27]:
# Подключаем все необходимые библиотеки
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
from scipy import stats as st
# Время обучения модели
import timeit
# Тренды и сезонность
from statsmodels.tsa.seasonal import seasonal_decompose
# Проверка на стационарность
from statsmodels.tsa.stattools import adfuller, kpss
# Проверка на дисперсию с помощью теста Андерсона-Дарлинга
from scipy.stats import anderson

from lightgbm import LGBMRegressor
# Вызов библиотеки для отключения предупреждения
import warnings

# Разбиение на обучающую, валидационную и тестовую выборку
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score, cross_val_predict
# Применим кроссвалидацию для повышения качеств обучения
# Для константной модели
from sklearn.dummy import DummyRegressor
from catboost import CatBoostRegressor

# Масштабируемость модели
from sklearn.preprocessing import StandardScaler, OrdinalEncoder, OneHotEncoder

# и для машинного обучения разными способами (по условию мы выбираем линейную регрессию):
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.tree import DecisionTreeRegressor

from sklearn.metrics import (
    # Точность модели
    accuracy_score,
    # Матрицы ошибок (для борьбы с дисбалансом)
    confusion_matrix, 
    # Полнота
    recall_score, 
    # Точность
    precision_score, 
    # F1-мера
    f1_score,
    # Метрика AUC-ROC
    roc_auc_score,
    roc_curve,
    # MSE
    mean_squared_error,
    mean_absolute_error,
    fbeta_score, 
    make_scorer
)

# Контроль выборки
from sklearn.utils import shuffle

### Загрузим данные

In [28]:
df = pd.read_csv('D:/Python/yandex_practicum/taxi.csv')

In [29]:
# Получаем датасет
try:
    # С локального файла
    try:
        # Сразу устанавливаем время как индексы
        df = pd.read_csv('taxi.csv', index_col=[0], parse_dates=[0])
    except:
        df = pd.read_csv('/datasets/taxi.csv', index_col=[0], parse_dates=[0])
except:
    print('Отсутствует датасет. Проверьте путь файла')

Отсутствует датасет. Проверьте путь файла


### Изучим данные

In [30]:
display(df.info())

display(df.describe())

display(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26496 entries, 0 to 26495
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   datetime    26496 non-null  object
 1   num_orders  26496 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 414.1+ KB


None

num_orders
count  26496.000000
mean      14.070463
std        9.211330
min        0.000000
25%        8.000000
50%       13.000000
75%       19.000000
max      119.000000

datetime  num_orders
0  2018-03-01 00:00:00           9
1  2018-03-01 00:10:00          14
2  2018-03-01 00:20:00          28
3  2018-03-01 00:30:00          20
4  2018-03-01 00:40:00          32

Видно, что средний заказ составляет `14`. А самым **максимальным** заказом был `119`. А сколько - за час, за день, за месяц, об этом нельзя сказать, пока не ресемплируем данные и сделаем анализ

### Предобработка данных

In [31]:
# Отсортируем данные на всякий
df.sort_index(inplace=True)

In [32]:
# Ресемплирируем данные по часами
df = df.resample('1H').sum()

TypeError: Only valid with DatetimeIndex, TimedeltaIndex or PeriodIndex, but got an instance of 'RangeIndex'

## Анализ

### Первый график временного ряда

In [ ]:
# Рассмотрим график и узнаем о нем
df.plot(figsize = (25,10))

Вполне нормальный график, виден медленный рост заказа. Однако есть сильный разброс, это вполне нормально, т.к. бывают праздничные дни, какие то мероприятия, пятничные дни и т.д.

### График со скользящими средними

In [ ]:
# Добавим скользящие среднее
df['rolling_mean'] = df.rolling(10).mean()

In [ ]:
# И снова рассмотрим график и узнаем о нем
df.plot(figsize=(20,10))

Здесь получилось неудобной и неинформативной, рассмотрим определенный промежуток времени, например, пусть промежуток времени с марта по апреля

In [ ]:
# И снова рассмотрим график и узнаем о нем
df['2018-03':'2018-04'].plot(figsize=(20,10))

### Исследуем стационарные ряды

In [ ]:
df['std'] = df['num_orders'].rolling(15).std()

In [ ]:
# И снова рассмотрим график и узнаем о нем
df['2018-03':'2018-04'].plot(figsize=(20,10))

In [ ]:
# Рассмотрим еще детально, пусть, за месяц
df['2018-03-01':'2018-04-01'].plot(figsize=(20,10))

*Примечание*:
- **Стохастический процесс** - это случайная величина, у которой со временем меняется её распределение. У этой величины есть среднее и дисперсия, которые тоже меняются

Стохастический процесс стационарный (англ. stationary stochastic process), если его распределение со временем не меняется. Например, к такому процессу относятся периодические колебания значений

Если распределение меняется, то процесс называется нестационарным

А что насчет временный ряд, его можно назвать стационарным или нет - нельзя, пока не проверим на тренды, сезонность и остаток декомпозиции

### Тренды и сезонность

In [ ]:
decomposed = seasonal_decompose(df['num_orders'])

In [ ]:
# Тренд
decomposed.trend.plot(ax=plt.gca(), figsize=(20,10))

plt.title('Тренд')

Видно, что тренд в последнее время увеличивается, т.е. с июня больше заказов на такси поступают нежели остальных

In [ ]:
# Сезонность
decomposed.seasonal.plot(ax=plt.gca(), figsize=(20,10))
plt.title('Сезонность')

Это говорит о том, что заказ стабилен и в любом сезоне, т.е. заказы всегда и будут поступать в любом времени сезона

In [ ]:
# Сезонность
decomposed.resid.plot(ax=plt.gca(), figsize=(20,10))
plt.title('Остаток декомпозиции')

### Проверка на дисперсию математическим способом

Можно проверить дисперсию, применяя теста Андерсона-Дарлинга

In [ ]:
# x - временной ряд
# dist='norm' - ожидаемое распределение (нормальное)
result = anderson(x=df['num_orders'], dist='norm')

# Статистика теста
test_statistic = result.statistic

# Критические значения для разных уровней значимости (1%, 5%, 10%)
critical_values = result.critical_values

# Если статистика теста больше критического значения для выбранного уровня значимости, то дисперсия не является постоянной
if test_statistic > critical_values[2]:
    print("Дисперсия не является постоянной")
else:
    print("Дисперсия является постоянной")

Здесь можно сказать, что этот промежуток времени, т.е. с марта 2018 до конца, является нестанционарным, т.к. присутствует тренд, который стремится увеличиваться, это можно заметить с момента июля 2018. А также дисперсия не является постоянной. Проверим на станционарность математическим способом, что она скажет

### Проверка на станционарность математическим способом

#### Тест Дики-Фуллера

In [ ]:
print('='*100)
# тест Дики-Фуллера на стационарность
print('Тест Дики-Фуллера на стационарность')
print('='*100)
result = adfuller(df['num_orders'])

print('ADF Статистка: %f' % result[0])
print('p-value: %f' % result[1])
print('Critical Values:')
for key, value in result[4].items():
    print('\t%s: %.3f' % (key, value))
    
print('='*100)

if(result[1] < 0.05):
    print('Принимаем альтернативную гипотезу. Временной ряд стационарен')
else:
    print('Отвергаем альтернативную гипотезу. Временной ряд не стационарен')
print('='*100)

#### Тест KPSS - Квятковский-Филлипс-Шмидт-ШинаДики-Фуллер 

In [ ]:
print('='*100)
# KPSS-тест на стационарность
# KPSS - Квятковского-Филлипса-Шмидта-Шина
print('KPSS-тест на стационарность')
print('='*100)
result = kpss(df['num_orders'])
print('KPSS Статистка: %f' % result[0])
print('p-value: %f' % result[1])
print('Critical Values:')
for key, value in result[3].items():
    print('\t%s: %.3f' % (key, value))
    
print('='*100)
#### Тест Дики-Фуллера
if(result[1] < 0.05):
    print('Принимаем альтернативную гипотезу. Временной ряд стационарен')
else:
    print('Отвергаем альтернативную гипотезу. Временной ряд не стационарен')
print('='*100)

2 разные теста одно и то же говорит, а по графику не сможем судить, что является стационарен. Если приглядеться, то можно увидеть небольшой рост, хоть капельку

## Обучение и тестирование

In [ ]:
# Здесь будем сохранить результаты машинного обучения
results = pd.DataFrame()

# А это будет счетчтиком для нумерация моеделй
count_model = 0

In [ ]:
# Создадим признаки, чтобы правильно обучить модель
def make_features(data, column, max_lag, rolling_mean_size):
    data['year'] = data.index.year
    data['month'] = data.index.month
    data['day'] = data.index.day
    data['dayofweek'] = data.index.dayofweek
    
    for lag in range(1, max_lag + 1):
        data['lag_{}'.format(lag)] = data[column].shift(lag)

    data['rolling_mean'] = data[column].shift().rolling(rolling_mean_size).mean()

### Деление на обучающей и тестовой выборки

In [78]:
# Создадим класс, который автоматически будет добиваться минимального RMSE
class search_small_rmse:
    # По умолчанию создается, если ничего не передаем аргументы
    def __init__(self, data):
        self.data = data
        self.rmse_v = 1_000_000
        self.max_lag = 1
        self.rolling_mean_size = 1
    
    # Создадим признаки, чтобы правильно обучить модель
    def make_features(self, data, column):
        data['year'] = data.index.year
        data['month'] = data.index.month
        data['day'] = data.index.day
        data['dayofweek'] = data.index.dayofweek

        for lag in range(1, self.max_lag + 1):
            data['lag_{}'.format(lag)] = data[column].shift(lag)

        data['rolling_mean'] = data[column].shift().rolling(self.rolling_mean_size).mean()

    # Разбиение на обучающие и тестовые выборки
    def make_train_test_split(self, name_features, name_target, test_size, random_state, shuffle=True):
        # Разделим обучающую и тестовую выборку
        self.train, self.test = train_test_split(self.data, test_size=test_size, random_state=random_state, shuffle=shuffle)
        self.train = self.train.dropna()
        
        self.features_train = self.train.drop(name_features, axis=1)
        self.target_train = self.train[name_target]

        self.features_test = self.test.drop(name_features, axis=1)
        self.target_test = self.test[name_target]
        
    # Здесь будет бесконечная итерация, которая добивается введенного желаемого RMSE
    def striving_rmse(self, name_model, parameters, cv_size, name_scoring, desired_rmse, max_value_lag, max_value_rol_mean_size):
        self.name_model = name_model
        self.parameters = parameters
        self.cv_size = cv_size
        self.name_scoring = name_scoring
        self.desired_rmse = desired_rmse
        self.max_value_lag = max_value_lag
        self.max_value_rol_mean_size = max_value_rol_mean_size


        for max_lag in range(1, max_value_lag):
            for rolling_mean_size in range(1, max_value_rol_mean_size):
                self.make_features(self.data, self.target_train)

                # Инициализируем модель
                self.model = GridSearchCV(name_model, param_grid = parameters, cv = cv_size, scoring = name_scoring)

                self.learning_model(self.features_train, self.target_train)
                self.predicting_model(self.features_test)

                if(self.rmse_v > self.desired_rmse):
                    print('Successfully')
                    break

    
    # Обучим модель на обучающей выборке
    def learning_model(self, features_train, target_train):
        self.model.fit(features_train, target_train)
        self.rmse_t = -model.best_score_
        self.refit_time_t = -model.refit_time_

    # Получим предсказания на тестовой выборки
    def predicting_model(self, features_test):
        self.features_test = features_test
        
        start_time = timeit.default_timer()
        
        self.predictions = self.model.predict(self.features_test)

        self.elapsed = round(timeit.default_timer() - start_time, 3)

        self.rmse_v = mean_squared_error(target_test, predictions, squared=False)

In [34]:
# На всякий случай сделаем копию, чтобы было удобно организовать несколько подборов
df_copy = df.copy()

In [83]:
df_class = search_small_rmse(df_copy)

In [85]:
df_class.make_train_test_split(name_features='num_orders', name_target='num_orders', test_size=0.1, random_state=12345, shuffle=False)

In [87]:
df_class.striving_rmse(name_model=LinearRegression(), parameters={}, cv_size=5, name_scoring='neg_root_mean_squared_error', desired_rmse=48, max_value_lag=24, max_value_rol_mean_size=48)

AttributeError: 'RangeIndex' object has no attribute 'year'

In [555]:
# На всякий случай сделаем копию, чтобы было удобно организовать несколько подборов
df_copy = df.copy()

In [556]:
# Функция для создания новых признаков
make_features(df_copy,'num_orders', 48, 96)

In [557]:
# Разделим обучающую и тестовую выборку
train, test = train_test_split(df_copy, test_size=(1 - 0.9), random_state=12345, shuffle=False)
train = train.dropna()

In [558]:
features_train = train.drop(['num_orders', 'std'], axis=1)
target_train = train['num_orders']

features_test = test.drop(['num_orders', 'std'], axis=1)
target_test = test['num_orders']

In [559]:
# Проверим
print(features_train.shape)
print(features_test.shape)

(3878, 53)
(442, 53)


In [560]:
print('Размер обучающей выборки:', round(features_train.shape[0] / df_copy.shape[0], 2), '%')
print('Размер тестовой выборки:', round(features_test.shape[0] / df_copy.shape[0], 2), '%')

Размер обучающей выборки: 0.88 %
Размер тестовой выборки: 0.1 %


In [561]:
# Проверим на пропущенные значения
print('-'*100)
print('features_train')
display(features_train.isna().sum())
print('-'*100)
print('features_test')
display(features_test.isna().sum())
print('-'*100)

----------------------------------------------------------------------------------------------------
features_train


rolling_mean    0
year            0
month           0
day             0
dayofweek       0
lag_1           0
lag_2           0
lag_3           0
lag_4           0
lag_5           0
lag_6           0
lag_7           0
lag_8           0
lag_9           0
lag_10          0
lag_11          0
lag_12          0
lag_13          0
lag_14          0
lag_15          0
lag_16          0
lag_17          0
lag_18          0
lag_19          0
lag_20          0
lag_21          0
lag_22          0
lag_23          0
lag_24          0
lag_25          0
lag_26          0
lag_27          0
lag_28          0
lag_29          0
lag_30          0
lag_31          0
lag_32          0
lag_33          0
lag_34          0
lag_35          0
lag_36          0
lag_37          0
lag_38          0
lag_39          0
lag_40          0
lag_41          0
lag_42          0
lag_43          0
lag_44          0
lag_45          0
lag_46          0
lag_47          0
lag_48          0
dtype: int64

----------------------------------------------------------------------------------------------------
features_test


rolling_mean    0
year            0
month           0
day             0
dayofweek       0
lag_1           0
lag_2           0
lag_3           0
lag_4           0
lag_5           0
lag_6           0
lag_7           0
lag_8           0
lag_9           0
lag_10          0
lag_11          0
lag_12          0
lag_13          0
lag_14          0
lag_15          0
lag_16          0
lag_17          0
lag_18          0
lag_19          0
lag_20          0
lag_21          0
lag_22          0
lag_23          0
lag_24          0
lag_25          0
lag_26          0
lag_27          0
lag_28          0
lag_29          0
lag_30          0
lag_31          0
lag_32          0
lag_33          0
lag_34          0
lag_35          0
lag_36          0
lag_37          0
lag_38          0
lag_39          0
lag_40          0
lag_41          0
lag_42          0
lag_43          0
lag_44          0
lag_45          0
lag_46          0
lag_47          0
lag_48          0
dtype: int64

----------------------------------------------------------------------------------------------------


Нет пропущенных значений, приступаем дальше

### LinearRegression

In [562]:
# Устанавливаем нужные параметры
parameters = {}

# Инициализируем модель
model = GridSearchCV(LinearRegression(), param_grid = parameters, cv = 5, scoring = 'neg_root_mean_squared_error')

In [563]:
%%notify -m "LinearRegression"
%%time

# Обучим модель на обучающей выборке
model.fit(features_train, target_train)

# Узнаем RMSE обучающей выборки
result_RMSE_t = -model.best_score_
print('RMSE TRAIN:', result_RMSE_t)
print('TIME TRAIN [s]:', round(time, 3))

RMSE TRAIN: 26.96991304685053
TIME TRAIN [s]: 0.348
CPU times: total: 46.9 ms
Wall time: 139 ms


<IPython.core.display.Javascript object>

**Проверка на тестовой выборки**

In [564]:
%%time
start_time = timeit.default_timer()
# Получим предсказания на тестовой выборки
predictions = model.predict(features_test)

elapsed = round(timeit.default_timer() - start_time, 3)

# Узнаем RMSE
result_RMSE_v = mean_squared_error(target_test, predictions, squared=False)
print('RMSE VALID:', result_RMSE_v)
print('Предсказание:', predictions.mean())

RMSE VALID: 43.75931704282299
Предсказание: 130.85972850678732
CPU times: total: 0 ns
Wall time: 4 ms


In [565]:
# Зафиксируем результаты
results[count_model] = pd.Series({
    'NAME': 'LinearRegression', 
    'RMSE TRAIN': result_RMSE_t, 
    'RMSE VALID': result_RMSE_v, 
    'PREDICTIONS': predictions.mean(), 
    'TIME TRAINING [s]': model.refit_time_, 
    'TIME PREDICTION [s]': elapsed, 
    'PARAMETRS': model.best_params_
})

results[count_model]
count_model+=1

### DecisionTreeRegressor

In [566]:
# Матрица гиперпараметров
parameters = {'max_depth': range(1, 201, 10), 'min_samples_leaf': [1, 0.5, 2]}

# Инициализируем модель c параметрами
model = GridSearchCV(DecisionTreeRegressor(random_state = 12345), param_grid = parameters, cv = 5, scoring = 'neg_root_mean_squared_error')

In [567]:
%%notify -m "DecisionTreeRegressor"
# %%time

# Обучим модель на обучающей выборке
model.fit(features_train, target_train)

# Узнаем RMSE обучающей выборки
result_RMSE_t = -model.best_score_
print('RMSE TRAIN:', result_RMSE_t)
print('TIME TRAIN [s]:', round(time, 2))

RMSE TRAIN: 31.497007061815196
TIME TRAIN [s]: 0.35


<IPython.core.display.Javascript object>

**Проверка на тестовой выборки**

In [568]:
%%time
start_time = timeit.default_timer()
# Получим предсказания на тестовой выборки
predictions = model.predict(features_test)

elapsed = round(timeit.default_timer() - start_time, 3)

result_RMSE_v = mean_squared_error(target_test, predictions, squared=False)
print('RMSE VALID:', result_RMSE_v)

print('Предсказание:', predictions.mean())

RMSE VALID: 56.57840477206501
Предсказание: 134.90192555661753
CPU times: total: 0 ns
Wall time: 4 ms


In [569]:
# Зафиксируем результаты
results[count_model] = pd.Series({
    'NAME': 'DecisionTreeRegressor', 
    'RMSE TRAIN': result_RMSE_t, 
    'RMSE VALID': result_RMSE_v, 
    'PREDICTIONS': predictions.mean(), 
    'TIME TRAINING [s]': model.refit_time_, 
    'TIME PREDICTION [s]': elapsed, 
    'PARAMETRS': model.best_params_
})

results[count_model]
count_model+=1

### RandomForestRegressor

In [570]:
# Матрица гиперпараметров для случайного леса
parameters = {'bootstrap': [True], 'max_depth': [5, 15], 'max_features': ['auto', 'log2'], 'n_estimators': [25, 50]}

# Инициализируем модель c параметрами
model = GridSearchCV(RandomForestRegressor(random_state = 12345), parameters, cv = 5, scoring = 'neg_root_mean_squared_error')

In [571]:
%%notify -m "RandomForestRegressor"
%%time

# Обучим модель на обучающей выборке
model.fit(features_train, target_train)

# Узнаем RMSE обучающей выборки
result_RMSE_t = -model.best_score_
print('RMSE TRAIN:', result_RMSE_t)
print('TIME TRAIN [s]:', round(time, 2))

D:\Program File\Anaconda\envs\practicum\lib\site-packages\sklearn\ensemble\_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(
D:\Program File\Anaconda\envs\practicum\lib\site-packages\sklearn\ensemble\_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(
D:\Program File\Anaconda\envs\practicum\lib\site-packages\sklearn\ensemble\_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as i

RMSE TRAIN: 23.55958374175849
TIME TRAIN [s]: 0.35
CPU times: total: 53.2 s
Wall time: 54.6 s


<IPython.core.display.Javascript object>

In [572]:
%%time
start_time = timeit.default_timer()
# Получим предсказания на тестовой выборки
predictions = model.predict(features_test)

elapsed = round(timeit.default_timer() - start_time, 3)

result_RMSE_v = mean_squared_error(target_test, predictions, squared=False)
print('RMSE VALID:', result_RMSE_v)

print('Предсказание:', predictions.mean())

RMSE VALID: 41.92059449252377
Предсказание: 146.98700083515004
CPU times: total: 0 ns
Wall time: 8.51 ms


In [573]:
# Зафиксируем результаты
results[count_model] = pd.Series({
    'NAME': 'RandomForestRegressor', 
    'RMSE TRAIN': result_RMSE_t, 
    'RMSE VALID': result_RMSE_v, 
    'PREDICTIONS': predictions.mean(), 
    'TIME TRAINING [s]': model.refit_time_, 
    'TIME PREDICTION [s]': elapsed, 
    'PARAMETRS': model.best_params_
})

results[count_model]
count_model+=1

### LightGBMRegressor

In [574]:
# Матрица гиперпараметров для LightGBMRegressor
parameters = {'num_leaves': [5, 10],
              'learning_rate': [0.1, 0.3],
              'max_depth': [3, 5],
              'n_estimators': [10, 25]}

# Инициализируем модель c параметрами
model = GridSearchCV(LGBMRegressor(), parameters, cv = 5, scoring = 'neg_root_mean_squared_error')

In [575]:
%%notify -m "LightGBMRegressor"
%%time

# Обучим модель на обучающей выборке
model.fit(features_train, target_train)

# Узнаем RMSE обучающей выборки
result_RMSE_t = -model.best_score_
print('RMSE TRAIN:', result_RMSE_t)
print('TIME TRAIN [s]:', round(time, 2))

RMSE TRAIN: 23.751787533341865
TIME TRAIN [s]: 0.35
CPU times: total: 23.8 s
Wall time: 1.94 s


<IPython.core.display.Javascript object>

**Проверка на тестовой выборки**

In [576]:
%%time
start_time = timeit.default_timer()
# Получим предсказания на тестовой выборки
predictions = model.predict(features_test)

elapsed = round(timeit.default_timer() - start_time, 3)

result_RMSE_v = mean_squared_error(target_test, predictions, squared=False)
print('RMSE VALID:', result_RMSE_v)
print('Предсказание:', predictions.mean())

RMSE VALID: 40.83300839480757
Предсказание: 135.53859315034362
CPU times: total: 0 ns
Wall time: 3 ms


In [577]:
# Зафиксируем результаты
results[count_model] = pd.Series({
    'NAME': 'LightGBMRegressor', 
    'RMSE TRAIN': result_RMSE_t, 
    'RMSE VALID': result_RMSE_v, 
    'PREDICTIONS': predictions.mean(), 
    'TIME TRAINING [s]': model.refit_time_, 
    'TIME PREDICTION [s]': elapsed, 
    'PARAMETRS': model.best_params_
})

results[count_model]
count_model+=1

### CatBoostRegressor

In [578]:
# Матрица гиперпараметров для LightGBMRegressor
parameters = {'depth' : [6, 8, 10],
              'learning_rate' : [0.01, 0.05, 0.1],
              'iterations' : [10, 15, 20]}

# Инициализируем модель c параметрами
model = GridSearchCV(CatBoostRegressor(), parameters, cv = 5, scoring = 'neg_root_mean_squared_error')

In [579]:
%%notify -m "CatBoostRegressor"
%%time

# Обучим модель на обучающей выборке
model.fit(features_train, target_train)

# Узнаем RMSE обучающей выборки
result_RMSE_t = -model.best_score_
print('RMSE TRAIN:', result_RMSE_t)
print('TIME TRAIN [s]:', round(time, 2))

0:	learn: 38.6610350	total: 5.33ms	remaining: 48ms
1:	learn: 38.4833528	total: 9.61ms	remaining: 38.4ms
2:	learn: 38.3164951	total: 15.2ms	remaining: 35.5ms
3:	learn: 38.1384116	total: 20.2ms	remaining: 30.3ms
4:	learn: 37.9708779	total: 25.4ms	remaining: 25.4ms
5:	learn: 37.7972442	total: 30.2ms	remaining: 20.1ms
6:	learn: 37.6453582	total: 34.9ms	remaining: 15ms
7:	learn: 37.4795710	total: 39.5ms	remaining: 9.87ms
8:	learn: 37.3273989	total: 44.4ms	remaining: 4.93ms
9:	learn: 37.1678681	total: 49.3ms	remaining: 0us
0:	learn: 39.2805171	total: 5.31ms	remaining: 47.8ms
1:	learn: 39.0908472	total: 9.95ms	remaining: 39.8ms
2:	learn: 38.9100834	total: 15.3ms	remaining: 35.8ms
3:	learn: 38.7281044	total: 20.8ms	remaining: 31.2ms
4:	learn: 38.5490087	total: 26ms	remaining: 26ms
5:	learn: 38.3561046	total: 30.9ms	remaining: 20.6ms
6:	learn: 38.1834601	total: 35ms	remaining: 15ms
7:	learn: 38.0167046	total: 39ms	remaining: 9.74ms
8:	learn: 37.8507677	total: 43.6ms	remaining: 4.84ms
9:	learn: 

0:	learn: 39.2805171	total: 4.21ms	remaining: 58.9ms
1:	learn: 39.0908472	total: 8.03ms	remaining: 52.2ms
2:	learn: 38.9100834	total: 11.9ms	remaining: 47.8ms
3:	learn: 38.7281044	total: 17.4ms	remaining: 47.8ms
4:	learn: 38.5490087	total: 21.3ms	remaining: 42.5ms
5:	learn: 38.3561046	total: 25.2ms	remaining: 37.7ms
6:	learn: 38.1834601	total: 29.4ms	remaining: 33.6ms
7:	learn: 38.0167046	total: 32.9ms	remaining: 28.8ms
8:	learn: 37.8507677	total: 37.1ms	remaining: 24.7ms
9:	learn: 37.6781016	total: 40.9ms	remaining: 20.4ms
10:	learn: 37.5089817	total: 44.9ms	remaining: 16.3ms
11:	learn: 37.3588963	total: 48.8ms	remaining: 12.2ms
12:	learn: 37.1967594	total: 52.6ms	remaining: 8.09ms
13:	learn: 37.0235882	total: 57ms	remaining: 4.07ms
14:	learn: 36.8704736	total: 60.8ms	remaining: 0us
0:	learn: 39.5130210	total: 3.54ms	remaining: 49.6ms
1:	learn: 39.3137123	total: 7.17ms	remaining: 46.6ms
2:	learn: 39.1275206	total: 10.7ms	remaining: 42.9ms
3:	learn: 38.9398545	total: 14.8ms	remaining: 

10:	learn: 27.6684210	total: 39.6ms	remaining: 14.4ms
11:	learn: 27.1669189	total: 43.8ms	remaining: 10.9ms
12:	learn: 26.7488798	total: 47.5ms	remaining: 7.31ms
13:	learn: 26.4001163	total: 51.2ms	remaining: 3.66ms
14:	learn: 26.0472826	total: 55.3ms	remaining: 0us
0:	learn: 37.7667111	total: 3.55ms	remaining: 49.7ms
1:	learn: 35.9997166	total: 6.98ms	remaining: 45.4ms
2:	learn: 34.5300329	total: 10.6ms	remaining: 42.5ms
3:	learn: 33.2745422	total: 15.1ms	remaining: 41.4ms
4:	learn: 32.0892241	total: 19.2ms	remaining: 38.4ms
5:	learn: 30.9962101	total: 23.2ms	remaining: 34.8ms
6:	learn: 30.2018846	total: 27ms	remaining: 30.9ms
7:	learn: 29.3982064	total: 31.2ms	remaining: 27.3ms
8:	learn: 28.6614764	total: 35.7ms	remaining: 23.8ms
9:	learn: 28.0158678	total: 39.7ms	remaining: 19.8ms
10:	learn: 27.5258964	total: 43.1ms	remaining: 15.7ms
11:	learn: 27.0188408	total: 46.7ms	remaining: 11.7ms
12:	learn: 26.5751161	total: 50.4ms	remaining: 7.76ms
13:	learn: 26.1940051	total: 54.3ms	remaini

0:	learn: 38.7266529	total: 4.69ms	remaining: 89.1ms
1:	learn: 37.7747290	total: 9.05ms	remaining: 81.5ms
2:	learn: 36.9332407	total: 13.4ms	remaining: 76.1ms
3:	learn: 36.1246512	total: 17.7ms	remaining: 70.7ms
4:	learn: 35.3687629	total: 22ms	remaining: 66ms
5:	learn: 34.6135981	total: 26.2ms	remaining: 61.1ms
6:	learn: 33.9733346	total: 30.9ms	remaining: 57.4ms
7:	learn: 33.3033876	total: 35.3ms	remaining: 53ms
8:	learn: 32.7775605	total: 39.5ms	remaining: 48.3ms
9:	learn: 32.2160079	total: 43.3ms	remaining: 43.3ms
10:	learn: 31.6531774	total: 46.7ms	remaining: 38.2ms
11:	learn: 31.1806603	total: 51.1ms	remaining: 34.1ms
12:	learn: 30.6825016	total: 55.4ms	remaining: 29.8ms
13:	learn: 30.2563126	total: 59.9ms	remaining: 25.7ms
14:	learn: 29.8386516	total: 63.9ms	remaining: 21.3ms
15:	learn: 29.4335623	total: 67.7ms	remaining: 16.9ms
16:	learn: 29.0827643	total: 71.7ms	remaining: 12.6ms
17:	learn: 28.7323682	total: 75.2ms	remaining: 8.36ms
18:	learn: 28.4262014	total: 79.1ms	remainin

0:	learn: 38.6521464	total: 13.6ms	remaining: 122ms
1:	learn: 38.4766847	total: 25.9ms	remaining: 104ms
2:	learn: 38.3131486	total: 37.5ms	remaining: 87.6ms
3:	learn: 38.1513043	total: 50.9ms	remaining: 76.3ms
4:	learn: 37.9743042	total: 63.5ms	remaining: 63.5ms
5:	learn: 37.7990968	total: 77ms	remaining: 51.3ms
6:	learn: 37.6239908	total: 89.1ms	remaining: 38.2ms
7:	learn: 37.4563867	total: 102ms	remaining: 25.5ms
8:	learn: 37.2836278	total: 114ms	remaining: 12.7ms
9:	learn: 37.1177689	total: 127ms	remaining: 0us
0:	learn: 39.2652540	total: 13.6ms	remaining: 123ms
1:	learn: 39.0745382	total: 26.7ms	remaining: 107ms
2:	learn: 38.8934555	total: 39ms	remaining: 90.9ms
3:	learn: 38.7036444	total: 51.5ms	remaining: 77.2ms
4:	learn: 38.5104004	total: 64.5ms	remaining: 64.5ms
5:	learn: 38.3389574	total: 76.9ms	remaining: 51.3ms
6:	learn: 38.1555598	total: 88.5ms	remaining: 37.9ms
7:	learn: 37.9739115	total: 101ms	remaining: 25.3ms
8:	learn: 37.7867102	total: 113ms	remaining: 12.6ms
9:	learn:

6:	learn: 38.1555598	total: 82.3ms	remaining: 94.1ms
7:	learn: 37.9739115	total: 94.6ms	remaining: 82.8ms
8:	learn: 37.7867102	total: 106ms	remaining: 70.4ms
9:	learn: 37.6049142	total: 118ms	remaining: 58.8ms
10:	learn: 37.4241459	total: 129ms	remaining: 46.9ms
11:	learn: 37.2492851	total: 141ms	remaining: 35.2ms
12:	learn: 37.0786676	total: 152ms	remaining: 23.4ms
13:	learn: 36.9058321	total: 177ms	remaining: 12.7ms
14:	learn: 36.7468582	total: 189ms	remaining: 0us
0:	learn: 39.5110438	total: 11.8ms	remaining: 165ms
1:	learn: 39.3161212	total: 23.3ms	remaining: 151ms
2:	learn: 39.1304877	total: 34.3ms	remaining: 137ms
3:	learn: 38.9436614	total: 45.9ms	remaining: 126ms
4:	learn: 38.7469916	total: 57.1ms	remaining: 114ms
5:	learn: 38.5525791	total: 68.8ms	remaining: 103ms
6:	learn: 38.3670086	total: 80.5ms	remaining: 92ms
7:	learn: 38.1760429	total: 92.1ms	remaining: 80.6ms
8:	learn: 37.9933596	total: 104ms	remaining: 69.1ms
9:	learn: 37.8094541	total: 115ms	remaining: 57.5ms
10:	lear

12:	learn: 25.8218056	total: 151ms	remaining: 23.2ms
13:	learn: 25.4089851	total: 162ms	remaining: 11.6ms
14:	learn: 24.9960430	total: 174ms	remaining: 0us
0:	learn: 37.5595434	total: 12.4ms	remaining: 174ms
1:	learn: 35.7254603	total: 23.1ms	remaining: 150ms
2:	learn: 34.2434237	total: 34.3ms	remaining: 137ms
3:	learn: 32.8286170	total: 45.6ms	remaining: 126ms
4:	learn: 31.6535860	total: 57.9ms	remaining: 116ms
5:	learn: 30.4460748	total: 69.6ms	remaining: 104ms
6:	learn: 29.4756239	total: 81.1ms	remaining: 92.7ms
7:	learn: 28.6236633	total: 93.6ms	remaining: 81.9ms
8:	learn: 27.8772743	total: 105ms	remaining: 70.1ms
9:	learn: 27.1500941	total: 117ms	remaining: 58.7ms
10:	learn: 26.5940190	total: 129ms	remaining: 46.9ms
11:	learn: 26.0048947	total: 140ms	remaining: 35ms
12:	learn: 25.5335642	total: 152ms	remaining: 23.4ms
13:	learn: 25.0629382	total: 164ms	remaining: 11.7ms
14:	learn: 24.6235524	total: 175ms	remaining: 0us
0:	learn: 32.7027296	total: 11.5ms	remaining: 161ms
1:	learn: 

0:	learn: 38.7158378	total: 13.1ms	remaining: 248ms
1:	learn: 37.7968375	total: 23.9ms	remaining: 215ms
2:	learn: 36.9603911	total: 35.4ms	remaining: 201ms
3:	learn: 36.1492160	total: 46.9ms	remaining: 188ms
4:	learn: 35.3332536	total: 57.9ms	remaining: 174ms
5:	learn: 34.5357494	total: 69.6ms	remaining: 162ms
6:	learn: 33.8252010	total: 81ms	remaining: 150ms
7:	learn: 33.1402831	total: 92.2ms	remaining: 138ms
8:	learn: 32.5210635	total: 103ms	remaining: 126ms
9:	learn: 31.9338877	total: 114ms	remaining: 114ms
10:	learn: 31.3931317	total: 126ms	remaining: 103ms
11:	learn: 30.8590755	total: 137ms	remaining: 91.3ms
12:	learn: 30.3536396	total: 148ms	remaining: 79.8ms
13:	learn: 29.9072336	total: 159ms	remaining: 68.2ms
14:	learn: 29.4557303	total: 171ms	remaining: 56.9ms
15:	learn: 29.0670562	total: 182ms	remaining: 45.4ms
16:	learn: 28.6789968	total: 193ms	remaining: 34ms
17:	learn: 28.3142383	total: 214ms	remaining: 23.8ms
18:	learn: 27.9550908	total: 225ms	remaining: 11.9ms
19:	learn:

0:	learn: 38.6508252	total: 58.5ms	remaining: 527ms
1:	learn: 38.4716508	total: 116ms	remaining: 466ms
2:	learn: 38.2938844	total: 183ms	remaining: 427ms
3:	learn: 38.1193178	total: 243ms	remaining: 365ms
4:	learn: 37.9451330	total: 301ms	remaining: 301ms
5:	learn: 37.7753822	total: 362ms	remaining: 241ms
6:	learn: 37.6043608	total: 420ms	remaining: 180ms
7:	learn: 37.4513468	total: 475ms	remaining: 119ms
8:	learn: 37.2910161	total: 534ms	remaining: 59.3ms
9:	learn: 37.1285860	total: 593ms	remaining: 0us
0:	learn: 39.2682622	total: 58.2ms	remaining: 524ms
1:	learn: 39.0722889	total: 119ms	remaining: 477ms
2:	learn: 38.8778945	total: 181ms	remaining: 422ms
3:	learn: 38.6954201	total: 247ms	remaining: 370ms
4:	learn: 38.5089268	total: 309ms	remaining: 309ms
5:	learn: 38.3235014	total: 368ms	remaining: 246ms
6:	learn: 38.1315937	total: 431ms	remaining: 185ms
7:	learn: 37.9629168	total: 494ms	remaining: 123ms
8:	learn: 37.7836757	total: 556ms	remaining: 61.8ms
9:	learn: 37.5992800	total: 6

12:	learn: 36.6602489	total: 801ms	remaining: 123ms
13:	learn: 36.5021905	total: 861ms	remaining: 61.5ms
14:	learn: 36.3432786	total: 920ms	remaining: 0us
0:	learn: 39.2682622	total: 64.6ms	remaining: 905ms
1:	learn: 39.0722889	total: 124ms	remaining: 806ms
2:	learn: 38.8778945	total: 189ms	remaining: 754ms
3:	learn: 38.6954201	total: 249ms	remaining: 685ms
4:	learn: 38.5089268	total: 310ms	remaining: 620ms
5:	learn: 38.3235014	total: 371ms	remaining: 556ms
6:	learn: 38.1315937	total: 427ms	remaining: 488ms
7:	learn: 37.9629168	total: 486ms	remaining: 425ms
8:	learn: 37.7836757	total: 541ms	remaining: 361ms
9:	learn: 37.5992800	total: 603ms	remaining: 302ms
10:	learn: 37.4265536	total: 678ms	remaining: 247ms
11:	learn: 37.2573020	total: 733ms	remaining: 183ms
12:	learn: 37.0770538	total: 791ms	remaining: 122ms
13:	learn: 36.9023983	total: 846ms	remaining: 60.4ms
14:	learn: 36.7255496	total: 902ms	remaining: 0us
0:	learn: 39.5121277	total: 63.9ms	remaining: 895ms
1:	learn: 39.3065704	to

7:	learn: 28.8574926	total: 453ms	remaining: 396ms
8:	learn: 28.0066064	total: 513ms	remaining: 342ms
9:	learn: 27.2304542	total: 568ms	remaining: 284ms
10:	learn: 26.5459002	total: 625ms	remaining: 227ms
11:	learn: 25.9640871	total: 684ms	remaining: 171ms
12:	learn: 25.4507387	total: 739ms	remaining: 114ms
13:	learn: 24.9061109	total: 806ms	remaining: 57.5ms
14:	learn: 24.4601577	total: 863ms	remaining: 0us
0:	learn: 37.7495305	total: 56ms	remaining: 784ms
1:	learn: 36.0055462	total: 112ms	remaining: 727ms
2:	learn: 34.5230255	total: 177ms	remaining: 707ms
3:	learn: 33.1627741	total: 232ms	remaining: 639ms
4:	learn: 31.8199458	total: 289ms	remaining: 577ms
5:	learn: 30.7189384	total: 344ms	remaining: 517ms
6:	learn: 29.6165926	total: 400ms	remaining: 457ms
7:	learn: 28.7838441	total: 465ms	remaining: 407ms
8:	learn: 27.9773207	total: 522ms	remaining: 348ms
9:	learn: 27.2185392	total: 581ms	remaining: 290ms
10:	learn: 26.4841301	total: 643ms	remaining: 234ms
11:	learn: 25.8673694	total

9:	learn: 31.7321918	total: 593ms	remaining: 593ms
10:	learn: 31.2432104	total: 650ms	remaining: 532ms
11:	learn: 30.7465541	total: 716ms	remaining: 477ms
12:	learn: 30.2610943	total: 773ms	remaining: 416ms
13:	learn: 29.7924820	total: 829ms	remaining: 355ms
14:	learn: 29.3496932	total: 883ms	remaining: 294ms
15:	learn: 28.9559371	total: 947ms	remaining: 237ms
16:	learn: 28.5516211	total: 1s	remaining: 177ms
17:	learn: 28.1885584	total: 1.06s	remaining: 118ms
18:	learn: 27.8662147	total: 1.11s	remaining: 58.7ms
19:	learn: 27.5317142	total: 1.18s	remaining: 0us
0:	learn: 38.5041556	total: 54.8ms	remaining: 1.04s
1:	learn: 37.5781774	total: 109ms	remaining: 980ms
2:	learn: 36.6828465	total: 163ms	remaining: 926ms
3:	learn: 35.8398210	total: 217ms	remaining: 870ms
4:	learn: 35.0457279	total: 281ms	remaining: 842ms
5:	learn: 34.3418039	total: 336ms	remaining: 783ms
6:	learn: 33.6405876	total: 389ms	remaining: 723ms
7:	learn: 33.0128757	total: 443ms	remaining: 665ms
8:	learn: 32.3795600	tot

12:	learn: 24.9672508	total: 699ms	remaining: 376ms
13:	learn: 24.4344489	total: 751ms	remaining: 322ms
14:	learn: 24.0048938	total: 803ms	remaining: 268ms
15:	learn: 23.4901735	total: 856ms	remaining: 214ms
16:	learn: 23.0390955	total: 917ms	remaining: 162ms
17:	learn: 22.6681255	total: 969ms	remaining: 108ms
18:	learn: 22.3170981	total: 1.02s	remaining: 53.8ms
19:	learn: 21.9310078	total: 1.07s	remaining: 0us
0:	learn: 32.7002893	total: 46.9ms	remaining: 891ms
1:	learn: 31.2899269	total: 99.7ms	remaining: 898ms
2:	learn: 30.0962467	total: 146ms	remaining: 830ms
3:	learn: 28.9721411	total: 193ms	remaining: 774ms
4:	learn: 27.8510683	total: 240ms	remaining: 721ms
5:	learn: 26.8756936	total: 289ms	remaining: 673ms
6:	learn: 26.0058536	total: 346ms	remaining: 642ms
7:	learn: 25.2855842	total: 394ms	remaining: 591ms
8:	learn: 24.5959585	total: 441ms	remaining: 539ms
9:	learn: 23.9539616	total: 488ms	remaining: 488ms
10:	learn: 23.3990703	total: 536ms	remaining: 439ms
11:	learn: 22.8397842

<IPython.core.display.Javascript object>

**Проверка на тестовой выборки**

In [580]:
%%time
start_time = timeit.default_timer()
# Получим предсказания на тестовой выборки
predictions = model.predict(features_test)

elapsed = round(timeit.default_timer() - start_time, 3)

result_RMSE_v = mean_squared_error(target_test, predictions, squared=False)
print('RMSE VALID:', result_RMSE_v)
print('Предсказание:', predictions.mean())

RMSE VALID: 52.466764457617806
Предсказание: 113.02367613909273
CPU times: total: 0 ns
Wall time: 4 ms


In [581]:
# Зафиксируем результаты
results[count_model] = pd.Series({
    'NAME': 'CatBoostRegressor', 
    'RMSE TRAIN': result_RMSE_t, 
    'RMSE VALID': result_RMSE_v, 
    'PREDICTIONS': predictions.mean(), 
    'TIME TRAINING [s]': model.refit_time_, 
    'TIME PREDICTION [s]': elapsed, 
    'PARAMETRS': model.best_params_
})

results[count_model]
count_model+=1

### Результаты

In [582]:
%%notify -m "Total result"
results = pd.DataFrame(results).T

<IPython.core.display.Javascript object>

In [583]:
# Смотрим на результаты
results

NAME RMSE TRAIN RMSE VALID PREDICTIONS TIME TRAINING [s]  \
0       LinearRegression  26.969913  43.759317  130.859729          0.021003   
1  DecisionTreeRegressor  31.497007  56.578405  134.901926          0.095559   
2  RandomForestRegressor  23.559584  41.920594  146.987001          5.414189   
3      LightGBMRegressor  23.751788  40.833008  135.538593          0.031538   
4      CatBoostRegressor  25.746336  52.466764  113.023676           0.26588   

  TIME PREDICTION [s]                                          PARAMETRS  
0               0.002                                                 {}  
1               0.002           {'max_depth': 11, 'min_samples_leaf': 1}  
2               0.008  {'bootstrap': True, 'max_depth': 15, 'max_feat...  
3               0.003  {'learning_rate': 0.3, 'max_depth': 5, 'n_esti...  
4               0.003  {'depth': 8, 'iterations': 20, 'learning_rate'...

In [584]:
# Нам интересует RMSE меньше 48, тогда
best_results = results.copy()
best_results = best_results[best_results['RMSE VALID'] <  48]

In [585]:
best_results.sort_values(by='RMSE VALID')

NAME RMSE TRAIN RMSE VALID PREDICTIONS TIME TRAINING [s]  \
3      LightGBMRegressor  23.751788  40.833008  135.538593          0.031538   
2  RandomForestRegressor  23.559584  41.920594  146.987001          5.414189   
0       LinearRegression  26.969913  43.759317  130.859729          0.021003   

  TIME PREDICTION [s]                                          PARAMETRS  
3               0.003  {'learning_rate': 0.3, 'max_depth': 5, 'n_esti...  
2               0.008  {'bootstrap': True, 'max_depth': 15, 'max_feat...  
0               0.002                                                 {}

## Вывод

Самой лучшей моделей является `LightGBMRegressor` - RMSE составляет **40.83**, а скорость обучения составляет почти 0,02 секунда, время предсказания - 0,2 секунда. Вполне очень акдеватный результат, чтобы выбрать эту модель для реализации. 

Есть другая модель - `RandomForestRegressor`, результаты неплохие, но недостатки у нее есть - скорость обучения и предсказания. 

А третье место заслужила простая модель - `LinearRegression`, однако эта модель может похвастаться скоростью обучения и предсказанией

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Данные проанализированы
- [x]  Модель обучена, гиперпараметры подобраны
- [x]  Качество моделей проверено, выводы сделаны
- [x]  Значение *RMSE* на тестовой выборке не больше 48